File: daily/Binary Search/1283-Find-the-smallest-divisor-given-a-threshold.py
from typing import List

class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        lo, hi = 1, max(nums)

        while lo < hi:
            mid = (lo + hi) // 2
            if self.valid_sum(nums, threshold, mid):
                hi = mid        # move LEFT → find first valid
            else:
                lo = mid + 1    # move RIGHT
        return lo

    def valid_sum(self, nums: List[int], threshold: int, divisor: int) -> bool:
        total = 0
        for num in nums:
            total += (num + divisor - 1) // divisor  # CEILING division
        return total <= threshold

File: daily/Binary Search/1283-Find-the-smallest-divisor-given-a-threshold.go
package daily

func smallestDivisor(nums []int, threshold int) int {
	maxNum := 0
	for _, val := range nums {
		if val > maxNum {
			maxNum = val
		}
	}

	ans := 0
	lo, hi := 1, maxNum
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if validSum(nums, threshold, mid) {
			ans = mid
			hi = mid - 1
		} else {
			lo = mid + 1
		}
	}
	return ans
}

func validSum(nums []int, threshold int, divisor int) bool {
	sum := 0
	for _, val := range nums {
		sum = sum + (val+divisor-1)/divisor
	}
	return sum <= threshold
}
